prusti-viper/src/encoder/backtranslation.rs

```diff
@@ -0,0 +1,457 @@
+use std::collections::HashMap;
+use std::convert::TryFrom;
+
+use viper::silicon_counterexample::*;
+use viper::VerificationError;
+use prusti_interface::data::ProcedureDefId;
+use prusti_common::vir::CfgMethod;
+use crate::encoder::Encoder;
+use crate::encoder::places::{Local, LocalVariableManager, Place};
+use crate::encoder::counterexample::*;
+
+use rustc_middle::mir;
+use rustc_middle::ty::{self, Ty, AdtKind, AdtDef, TyCtxt};
+use rustc_span::MultiSpan;
+
+
+pub struct Backtranslator<'tcx>{
+    mir: mir::Body<'tcx>,
+    def_id: ProcedureDefId,
+    silicon_counterexample: SiliconCounterexample,
+    tcx: TyCtxt<'tcx>,
+    is_pure: bool,
+    disc_info: HashMap<(ProcedureDefId, String), Vec<String>>,
+}
+
+impl<'tcx> Backtranslator<'tcx> {
+    pub fn new(
+        encoder: &Encoder<'_, 'tcx>,
+        def_id: ProcedureDefId,
+        silicon_opt_counterexample: Option<SiliconCounterexample>
```

Comment:
Maybe this shouldn't even be called when there is no counterexample, i.e. move the `let Some(...)` out?